internal/controller/memcached_controller.go

```diff
@@ -75,12 +76,106 @@ func (r *MemcachedReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 
 	log.Info("1. Fetch the memcached instance. Memcached resource found", "memcached.Name", memcached.Name, "memcached.Namespace", memcached.Namespace)
 
-	// 2. Check if deployment exists, if not create one
-	found := &appsv1.Deployment{}
-	err = r.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: req.Name}, found)
+	// 5a. Check if spec version is new, whether it is a minor upgrade, and write the new version to the status if yes.
+	version := memcached.Spec.Version
+	currentVersion := memcached.Status.CurrentVersion
```

Comment:
This seems very unconventional. The value here is being read from the status, a field the controller is supposed to write to for information flow to the outside world. I don't object to `checkedVersion`, as this can be seen as an optimization.

Reply:
I'm wondering about this. Is it really a misuse to read from the status for the controller's internal logic? In this case, it's to avoid parsing a string to extract a version, which is simple enough, but what if, hypothetically, the process of extracting the current state from "the real world" was difficult, expensive, or error prone? What would the right place be to store such a value?

Reply:
I wouldn't call it strictly forbidden to write read from the status, but it should be avoided. The status mainly exists so that users have an insight into how their requested resource is doing. The risk here is "state drift"(https://thenewstack.io/writing-custom-kubernetes-controllers-beware-of-state-drift/). The managed object (Deployment) might change without a change to the status, for example a user manually changes the image. In that case the `currentVersion` and the actual image version no longer match. Interactions like these can make it harder to reason about the code, and always need to be considered.

For example: the user manually updates the image of the deployment to a new major version manually, which can be upgraded automatically, but not downgraded without data loss. The current code walks through all cases of currentVersion, version, and  checkedVersion. They all match, so the status is not updated. 
Now we get to line 195, where the deployment is updated. The image does not match `currentVersion`, to the controller updates it. This constitutes a downgrade of the image, which is out of spec for the controller.

So in cases where it is difficult, expensive, or error prone this needs to be considered. "What's the risk of the value going out of date, and how meaningful is accuracy of the field to the operator and the user?". There may also be cases where there's no other way than to use the status as a source of truth.
There are also backup solutions for kubernetes that don't backup the status of a resource by default, another issue to be considered.

Reply:
Thank you! I didn't consider the image in the deployment changing due to outside factors, which makes it clear that my logic doesn't work. Lesson learned: don't try to wrangle with state.